homeassistant/components/gpslogger/__init__.py

```diff
@@ -5,110 +5,116 @@
 https://home-assistant.io/components/gpslogger/
 """
 import logging
-from hmac import compare_digest
 
 import voluptuous as vol
-from aiohttp.web_exceptions import HTTPUnauthorized
-from aiohttp.web_request import Request
+from aiohttp import web
 
 import homeassistant.helpers.config_validation as cv
-from homeassistant.components.http import HomeAssistantView, CONF_API_PASSWORD
-from homeassistant.const import CONF_PASSWORD, HTTP_UNPROCESSABLE_ENTITY
+from homeassistant.components.device_tracker import ATTR_BATTERY
+from homeassistant.components.device_tracker.tile import ATTR_ALTITUDE
+from homeassistant.const import HTTP_UNPROCESSABLE_ENTITY, \
+    HTTP_OK, ATTR_LATITUDE, ATTR_LONGITUDE, CONF_WEBHOOK_ID
+from homeassistant.helpers import config_entry_flow
 from homeassistant.helpers.discovery import async_load_platform
 from homeassistant.helpers.dispatcher import async_dispatcher_send
 
 _LOGGER = logging.getLogger(__name__)
 
 DOMAIN = 'gpslogger'
-DEPENDENCIES = ['http']
+DEPENDENCIES = ['webhook']
 
-CONFIG_SCHEMA = vol.Schema({
-    vol.Optional(DOMAIN): vol.Schema({
-        vol.Optional(CONF_PASSWORD): cv.string,
-    }),
-}, extra=vol.ALLOW_EXTRA)
+TRACKER_UPDATE = '{}_tracker_update'.format(DOMAIN)
 
-URL = '/api/{}'.format(DOMAIN)
+ATTR_ACCURACY = 'accuracy'
+ATTR_ACTIVITY = 'activity'
+ATTR_DEVICE = 'device'
+ATTR_DIRECTION = 'direction'
+ATTR_PROVIDER = 'provider'
+ATTR_SPEED = 'speed'
+
+DEFAULT_ACCURACY = 200
+DEFAULT_BATTERY = -1
+
+
+def _id(value: str) -> str:
+    """Coerce id by removing '-'."""
+    return value.replace('-', '')
 
-TRACKER_UPDATE = '{}_tracker_update'.format(DOMAIN)
+
+WEBHOOK_SCHEMA = vol.Schema({
+    vol.Required(ATTR_LATITUDE): cv.latitude,
+    vol.Required(ATTR_LONGITUDE): cv.longitude,
+    vol.Required(ATTR_DEVICE): _id,
+    vol.Optional(ATTR_ACCURACY, default=DEFAULT_ACCURACY): vol.Coerce(float),
+    vol.Optional(ATTR_BATTERY, default=DEFAULT_BATTERY): vol.Coerce(float),
+    vol.Optional(ATTR_SPEED): vol.Coerce(float),
+    vol.Optional(ATTR_DIRECTION): vol.Coerce(float),
+    vol.Optional(ATTR_ALTITUDE): vol.Coerce(float),
+    vol.Optional(ATTR_PROVIDER): cv.string,
+    vol.Optional(ATTR_ACTIVITY): cv.string
+})
 
 
 async def async_setup(hass, hass_config):
     """Set up the GPSLogger component."""
-    config = hass_config[DOMAIN]
-    hass.http.register_view(GPSLoggerView(config))
-
     hass.async_create_task(
         async_load_platform(hass, 'device_tracker', DOMAIN, {}, hass_config)
```

Comment:
Shouldn't we forward setup entry instead from `async_setup_entry` below?

Reply:
You're right.

Reply:
Sounds like something I missed with geofency, gpslogger and locative.

I'll put out PRs for all three of those. I think we can implement the teardown as well.

Reply:
So I'm a little confused at how this is supposed to be implemented. According to the [docs](https://developers.home-assistant.io/docs/en/config_entries_index.html), the platform needs to support the following method:

`async def async_setup_entry(hass, config_entry, async_add_devices):`

How do I get access to `async_see` from there?

Reply:
That's what I had tried and was not able to get working. Do you mind taking a look at this branch? https://github.com/rohankapoorcom/home-assistant/commit/bf6e61869c1a93b9c2ae05934042eeb84daea0c6

Error log attached. I'm not seeing any stack traces.
```
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.setup] Setting up gpslogger
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.setup] Setup of domain gpslogger took 0.0 seconds.
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.setup] Setup of domain config took 0.0 seconds.
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.loader] Loaded device_tracker from homeassistant.components.device_tracker
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.setup] Setting up device_tracker
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.setup] Setup of domain device_tracker took 0.0 seconds.
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.loader] Loaded device_tracker.gpslogger from homeassistant.components.device_tracker.gpslogger
2019-01-16 23:29:31 INFO (MainThread) [homeassistant.components.device_tracker] Setting up device_tracker.gpslogger
2019-01-16 23:29:31 ERROR (MainThread) [homeassistant.components.device_tracker] Error setting up platform gpslogger
```

Reply:
Remove `async_setup_scanner`. If it exists in the module it will take precedence over `async_setup_entry`.

Reply:
Got it, Thanks Martin.

Reply:
Pushed up in #20237.